File: typescript/src/agents/experimental/remote/agent.ts
/**
 * Copyright 2025 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Callback } from "@/emitter/types.js";
import { Emitter } from "@/emitter/emitter.js";
import { AgentError, BaseAgent, BaseAgentRunOptions } from "@/agents/base.js";
import { GetRunContext } from "@/context.js";
import { AssistantMessage, Message } from "@/backend/message.js";
import { BaseMemory } from "@/memory/base.js";
import { Client as MCPClient } from "@i-am-bee/acp-sdk/client/index.js";
import { Transport } from "@i-am-bee/acp-sdk/shared/transport.js";
import { shallowCopy } from "@/serializer/utils.js";
import { NotImplementedError } from "@/errors.js";
import { AgentRunProgressNotificationSchema } from "@i-am-bee/acp-sdk/types.js";
import { SSEClientTransport } from "@i-am-bee/acp-sdk/client/sse.js";

export interface RemoteAgentRunInput {
  prompt: Record<string, unknown> | string;
}

export interface RemoteAgentRunOutput {
  message: Message;
}

export interface RemoteAgentEvents {
  update: Callback<{ output: string }>;
}

interface Input {
  client: MCPClient;
  transport: Transport;
  agentName: string;
}

export class RemoteAgent extends BaseAgent<RemoteAgentRunInput, RemoteAgentRunOutput> {
  public emitter = Emitter.root.child<RemoteAgentEvents>({
    namespace: ["agent", "remote"],
    creator: this,
  });

  constructor(protected readonly input: Input) {
    super();
  }

  protected async _run(
    input: RemoteAgentRunInput,
    _options: BaseAgentRunOptions,
    context: GetRunContext<this>,
  ): Promise<RemoteAgentRunOutput> {
    const runner = this.createRunner(context);

    const output = await runner(input);

    const message: Message = new AssistantMessage(output);

    return { message };
  }

  protected async listAgents() {
    const response = await this.input.client.listAgents();

    return response.agents;
  }

  protected createRunner(context: GetRunContext<this>) {
    return async (input: RemoteAgentRunInput): Promise<string> => {
      try {
        this.input.client.setNotificationHandler(
          AgentRunProgressNotificationSchema,
          async (notification) => {
            await context.emitter.emit("update", {
              output: JSON.stringify(notification.params.delta, null, 2),
            });
          },
        );
        if (!this.input.client.transport) {
          await this.input.client.connect(this.input.transport);
        }
      } catch (e) {
        throw new AgentError(`Can't connect to Beeai Platform.`, [e], { isFatal: true });
      }

      const agents = await this.listAgents();
      const agent = agents.find((agent) => agent.name === this.input.agentName);
      if (!agent) {
        throw new AgentError(
          `Agent ${this.input.agentName} is not registered in the platform`,
          [],
          {
            isFatal: true,
          },
        );
      }

      const response = await this.input.client.runAgent(
        {
          name: this.input.agentName,
          input: typeof input.prompt === "string" ? { text: input.prompt } : input.prompt,
        },
        {
          timeout: 10_000_000,
          signal: context.signal,
          onprogress: () => null, // This has to be here in order for notifications to work.
        },
      );

      const output = JSON.stringify(response.output, null, 2);
      await context.emitter.emit("update", { output });
      return output;
    };
  }

  get memory() {
    throw new NotImplementedError();
  }

  set memory(memory: BaseMemory) {
    throw new NotImplementedError();
  }

  createSnapshot() {
    return {
      ...super.createSnapshot(),
      input: shallowCopy(this.input),
      emitter: this.emitter,
    };
  }

  static createSSEAgent(url: string, agentName: string) {
    return new RemoteAgent({
      client: new MCPClient({
        name: "remote-agent",
        version: "1.0.0",
      }),
      transport: new SSEClientTransport(new URL(url)),
      agentName: agentName,
    });
  }
}

File: python/beeai_framework/agents/experimental/remote/agent.py
# Copyright 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import uuid
from collections.abc import AsyncGenerator
from contextlib import AsyncExitStack
from typing import Any

import anyio
from acp import ClientSession, ServerNotification
from acp.client.sse import sse_client
from acp.shared.session import ReceiveResultT
from acp.types import (
    AgentRunProgressNotification,
    AgentRunProgressNotificationParams,
    ClientRequest,
    RequestParams,
    RunAgentRequest,
    RunAgentRequestParams,
    RunAgentResult,
)

from beeai_framework.agents.base import BaseAgent
from beeai_framework.agents.errors import AgentError
from beeai_framework.agents.experimental.remote.events import (
    RemoteAgentErrorEvent,
    RemoteAgentUpdateEvent,
    RemoteAgentWarningEvent,
    remote_agent_event_types,
)
from beeai_framework.agents.experimental.remote.types import (
    RemoteAgentInput,
    RemoteAgentRunOutput,
)
from beeai_framework.backend.message import AssistantMessage
from beeai_framework.cancellation import AbortSignal
from beeai_framework.context import Run, RunContext
from beeai_framework.emitter import Emitter
from beeai_framework.errors import FrameworkError
from beeai_framework.memory import BaseMemory


class RemoteAgent(BaseAgent[RemoteAgentRunOutput]):
    def __init__(self, agent_name: str, *, url: str) -> None:
        super().__init__()
        self.input = RemoteAgentInput(agent_name=agent_name, url=url)
        self.exit_stack = AsyncExitStack()

    def run(
        self,
        input: str | dict[str, Any],
        *,
        signal: AbortSignal | None = None,
    ) -> Run[RemoteAgentRunOutput]:
        async def handler(context: RunContext) -> RemoteAgentRunOutput:
            session: ClientSession = await self._connect_to_server()

            async with self.exit_stack:
                async for message in self._send_request_with_notifications(
                    context,
                    session,
                    client_req=RunAgentRequest(
                        method="agents/run",
                        params=RunAgentRequestParams(
                            name=self.input.agent_name, input={"text": input} if isinstance(input, str) else input
                        ),
                    ),
                    result_type=RunAgentResult,
                ):
                    match message:
                        case ServerNotification(
                            root=AgentRunProgressNotification(params=AgentRunProgressNotificationParams(delta=delta))
                        ):
                            await context.emitter.emit("update", RemoteAgentUpdateEvent(key="update", value=delta))
                        case RunAgentResult() as result:
                            await context.emitter.emit(
                                "update", RemoteAgentUpdateEvent(key="final_answer", value=result.output)
                            )
                            return RemoteAgentRunOutput(result=AssistantMessage(json.dumps(result.output)))

            raise AgentError("No response from assistant.")

        return self._to_run(
            handler,
            signal=signal,
            run_params={
                "prompt": input,
                "signal": signal,
            },
        )

    async def _send_request_with_notifications(
        self,
        context: RunContext,
        session: ClientSession,
        client_req: RunAgentRequest,
        result_type: type[ReceiveResultT],
    ) -> AsyncGenerator[ReceiveResultT | ServerNotification | None, None]:
        resp: ReceiveResultT | None = None
        async with AsyncExitStack():
            message_writer, message_reader = anyio.create_memory_object_stream[ServerNotification]()

            req_root = ClientRequest(client_req).root
            req_root.params = client_req.params or RequestParams()
            req_root.params.meta = RequestParams.Meta(progressToken=uuid.uuid4().hex)
            final_req = ClientRequest(req_root)

            async with anyio.create_task_group() as task_group:

                async def request_task() -> None:
                    nonlocal resp
                    try:
                        resp = await session.send_request(final_req, result_type)
                    except Exception as e:
                        e = FrameworkError.ensure(e)
                        await context.emitter.emit(
                            "error",
                            RemoteAgentErrorEvent(message="Unable to send request", error=e),
                        )
                    finally:
                        task_group.cancel_scope.cancel()

                async def read_notifications() -> None:
                    # IMPORTANT(!) if the client does not read the notifications, agent gets blocked
                    async for message in session.incoming_messages:
                        try:
                            if isinstance(message, Exception):
                                raise AgentError("Remote agent error", cause=message)
                            notification = ServerNotification.model_validate(message)
                            await message_writer.send(notification)
                        except ValueError as e:
                            await context.emitter.emit(
                                "warning",
                                RemoteAgentWarningEvent(
                                    message=f"Unable to parse message from server: {message}", data=e
                                ),
                            )

                task_group.start_soon(read_notifications)
                task_group.start_soon(request_task)

                async for message in message_reader:
                    yield message

        if resp:
            yield resp

    async def _connect_to_server(
        self,
    ) -> ClientSession:
        try:
            sse_transport = await self.exit_stack.enter_async_context(sse_client(url=self.input.url))
            self.read, self.write = sse_transport
            session: ClientSession = await self.exit_stack.enter_async_context(ClientSession(self.read, self.write))
            await session.initialize()
            response = await session.list_agents()
            agents = response.agents

            agent = any(agent.name == self.input.agent_name for agent in agents)
            if not agent:
                raise AgentError(f"Agent {self.input.agent_name} is not registered in the platform")
            return session

        except Exception as e:
            raise AgentError("Can't connect to Beeai Platform.", cause=e)

    def _create_emitter(self) -> Emitter:
        return Emitter.root().child(
            namespace=["agent", "remote"],
            creator=self,
            events=remote_agent_event_types,
        )

    @property
    def memory(self) -> BaseMemory:
        raise NotImplementedError()

    @memory.setter
    def memory(self, memory: BaseMemory) -> None:
        raise NotImplementedError()
